sdk/lib/atl/atlconv.h

```diff
@@ -0,0 +1,368 @@
+    {
+        if (!psz)
+        {
+            m_psz = NULL;
+            m_szBuffer[0] = 0;
+            return;
+        }
+        int cchMax = lstrlenW(psz);
+        if (cchMax <= t_nBufferLength)
+        {
+#ifdef _STRSAFE_H_INCLUDED_
+            StringCchCopyW(m_szBuffer, _countof(m_szBuffer), psz);
+#else
+            lstrcpynW(m_szBuffer, psz, _countof(m_szBuffer));
+#endif
+            m_psz = m_szBuffer;
+            return;
+        }
+
+        m_szBuffer[0] = 0;
+        m_psz = _wcsdup(psz);
+        if (!m_psz)
+            AtlThrow(E_OUTOFMEMORY);
+    }
+};
+
+template <int t_nBufferLength = 128>
+class CA2WEX
+{
+public:
+    LPWSTR m_psz;
+    wchar_t m_szBuffer[t_nBufferLength];
+
+    CA2WEX(_In_z_ LPCSTR psz)
+    {
+        Init(psz, CP_ACP);
+    }
+
+    CA2WEX(_In_z_ LPCSTR psz, _In_ UINT nCodePage)
+    {
+        Init(psz, nCodePage);
+    }
+
+    ~CA2WEX() throw()
+    {
+        if (m_psz != m_szBuffer)
+            free(m_psz);
+    }
+
+    _Ret_z_ operator LPWSTR() const throw()
+    {
+        return m_psz;
+    }
+
+private:
+    // CA2WEX is not copyable
+    CA2WEX(_In_ const CA2WEX&) throw() = delete;
+    CA2WEX& operator=(_In_ const CA2WEX&) throw() = delete;
+
+    void Init(_In_z_ LPCSTR psz, _In_ UINT nCodePage)
+    {
+        if (!psz)
+        {
+            m_psz = NULL;
+            m_szBuffer[0] = 0;
+            return;
+        }
+
+        // Calculation of MultiByteToWideChar is slow. Use lstrlenA instead.
+        int cchMax = lstrlenA(psz) + 1;
+        if (cchMax <= (int)_countof(m_szBuffer))
+        {
+            // Use the static buffer
+            m_psz = m_szBuffer;
+            cchMax = _countof(m_szBuffer);
+        }
+        else
+        {
+            // Allocate a new buffer
+            m_szBuffer[0] = 0;
+            m_psz = (LPWSTR)malloc(cchMax * sizeof(WCHAR));
+            if (!m_psz)
+                AtlThrow(E_OUTOFMEMORY);
+        }
+
+        MultiByteToWideChar(nCodePage, 0, psz, -1, m_psz, cchMax);
+        m_psz[cchMax - 1] = 0;
+    }
+};
+
+template <int t_nBufferLength = 128>
+class CW2AEX
+{
+public:
+    LPSTR m_psz;
+    char m_szBuffer[t_nBufferLength];
+
+    CW2AEX(_In_z_ LPCWSTR psz)
+    {
+        Init(psz, CP_ACP);
+    }
+
+    CW2AEX(_In_z_ LPCWSTR psz, _In_ UINT nCodePage)
+    {
+        Init(psz, nCodePage);
+    }
+
+    ~CW2AEX() throw()
+    {
+        if (m_psz != m_szBuffer)
+            free(m_psz);
+    }
+
+    _Ret_z_ operator LPSTR() const throw()
+    {
+        return m_psz;
+    }
+
+private:
+    // CW2AEX is not copyable
+    CW2AEX(_In_ const CW2AEX&) throw() = delete;
+    CW2AEX& operator=(_In_ const CW2AEX&) throw() = delete;
+
+    void Init(_In_z_ LPCWSTR psz, _In_ UINT nConvertCodePage)
+    {
+        if (!psz)
+        {
+            m_psz = NULL;
+            m_szBuffer[0] = 0;
+            return;
+        }
+
+        // Calculation of WideCharToMultiByte is slow. Use lstrlenW instead.
```

Comment:
Was this a problem while testing?
I'd prefer if there were no shortcuts like this unless it is actually needed, because increased complexity leads to more bugs.

Reply:
> Was this a problem while testing?

Yep. There's a corner case against this implementation. It looks like an optimization.

Reply:
The failure like this:
```txt
atlconv.cpp:125: Test failed: w2ax.m_psz == w2ax.m_szBuffer
```
happens.

Reply:
Some #ifdef's added. Please compare these codes.